clippy_lints/src/transmuting_null.rs

```diff
@@ -0,0 +1,116 @@
+use crate::utils::{match_def_path, match_qpath, opt_def_id, paths, span_lint};
+use crate::consts::{constant_context, Constant};
+use if_chain::if_chain;
+use rustc::lint::{LintArray, LintPass, LateLintPass, LateContext};
+use rustc::{declare_tool_lint, lint_array};
+use rustc::hir::{Expr, ExprKind};
+use syntax::ast::{LitKind};
+
+declare_clippy_lint! {
+    pub TRANSMUTING_NULL,
+    correctness,
+    "transmutes from a null pointer to a reference, which is Undefined Behavior"
+}
+
+#[derive(Copy, Clone)]
+pub struct TransmutingNullPass;
+
+impl LintPass for TransmutingNullPass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(
+            TRANSMUTING_NULL,
+        )
+    }
+
+    fn name(&self) -> &'static str {
+        "TransmutingNull"
+    }
+}
+
+const LINT_MSG : &str = "transmuting a known null pointer into a reference.";
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TransmutingNullPass {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        
+
+        // Catching one liner statements
+        // FIXME: investigate if it is possible to catch all 
+        // null-transmuting statements with this one if-chain.
+        if_chain! {
+            if let ExprKind::Call(ref func, ref args) = expr.node;
+            if let ExprKind::Path(ref path) = func.node;
+            if match_qpath(path, &["std", "mem", "transmute"]);
+            if args.len() == 1;
+            
+            then {
+
+                // Catching transmute over variables or constants 
+                // that resolve to `null`
+                // FIXME: make it work
+                let mut const_eval_context = constant_context(cx, cx.tables);
+                if_chain! {
+                    if let ExprKind::Path(ref qpath) = args[0].node;
+
+                    then {
+                        let x = const_eval_context.expr(&args[0]);
```

Comment:
You can put let bindings in the `if_chain!` macro:
```rust
if_chain! {
    if let ...;
    let x = 0;
    if x > 10;
    then {

    }
}
```

Reply:
What. That's awesome! This is such a nice macro, I showed it to another student who's getting into Rust and we're both in love with ir <3

> I will take another look tomorrow.

Thanks again for your help! I hope there's a way 🤞🏻🤞🏻

Reply:
This if_chain can still be collapsed.